### PR TITLE
LSIF: Guard all backend actions behind enterprise

### PR DESCRIPTION
**Pull request by efritz**
**Summary**: Ensure that LSIF-related resolvers and the lsif-server proxy is only registered when running in enterprise mode. See [this slack thread](https://sourcegraph.slack.com/archives/C0C324C91/p1573579387122800) for motivation.

**What changed**:

- GraphQL changes:
  - Move `lsif_*.go` from `graphqlbackend` to `enterprise/codeintelligence/resolvers`. This is the bulk of the change as it required updating imports and reshuffling some of the code. No logic changed *except* for the creation of the commit resolver for lsif dumps (see `LsifDumps` method in the resolver).
  - Add `graphqlbackend/code_intelligence.go`, which defines the interfaces for each resolver and updates the implementation of schemaResolver to call the optionally-bound `CodeIntelligenceResolver`
  - Update concrete uses of old resolvers in `graphqlbackend` to use the interfaces defined in the bullet point above.
  - Add `Type` method to `RepositoryResolver` so that we can get the repository db record from a resolver. I didn't see another way to do this and I was previously using package-private variables. **Feedback welcome on this point.**
  - Renamed variable `onlyInEnterprise` to `a8nOnlyInEnterprise` to avoid conflicts between a8n and codeintelligence resolvers. This keeps the text of the error relevant to the endpoint that emits it.
- HTTP changes:
  - Move the LSIF endpoints into `enterprise/codeintelligence/lsifserver/proxy`. 
  - Add a non-internal package in the frontend `httpapi` which only contains a constant for the LSIF server proxy struct. This is set by enterprise mode. This change was meant to reflect the same way resolvers are set in the enterprise frontend for graphqlbackend (there is both an internal and non-internal version).
  - Update LSIF route registration to use the variable described in the bullet point above. If that variable is not set, then register a dummy "404 feature not enabled" endpoint.

### CODE REVIEW

**Comment by unknwon** [comment]
This feels a lot context to carry on during the review, I'll do it tomorrow PST.

**Comment by efritz** [comment]
> This feels a lot context to carry on during the review

If you have a suggestion to break it up I'll certainly put in the effort!

**Comment by unknwon** [comment]
> If you have a suggestion to break it up I'll certainly put in the effort!

I guess this is what happens when refactoring. I'm just almost brain dead today 😝

**Comment by unknwon** [comment]
One little nit suggestion is does it make sense to rename `codeintelligence` to `codeintel`, because "codeintel" is kind of a short convention name?

File: cmd/frontend/internal/httpapi/httpapi.go [hunk 1]
```diff
@@ -62,13 +60,22 @@ func NewHandler(m *mux.Router, schema *graphql.Schema, githubWebhook http.Handle
 
 	m.Get(apirouter.GraphQL).Handler(trace.TraceRoute(handler(serveGraphQL(schema))))
 
-	lsifServerURL, err := url.Parse(lsif.ServerURLFromEnv)
-	if err != nil {
-		log15.Error("skipping initialization of the LSIF HTTP API because the environment variable LSIF_SERVER_URL is not a valid URL", "parse_error", err, "value", lsif.ServerURLFromEnv)
+	if httpapi.NewLSIFServerProxy != nil {
```
Comment (tsenart) [inline cmd/frontend/internal/httpapi/httpapi.go]: Any chance we could avoid now returning an error from this function for this single use case by parameterising the proxy as an argument (i.e. like `schema` and `githubWebhook`)?

File: enterprise/pkg/codeintel/lsifserver/proxy/auth.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+		return errors.New("Must provide github_token."), http.StatusUnauthorized
+	}
+
+	client := github.NewClient(&githubURL, githubToken, nil)
+	repo, err := client.GetRepository(r.Context(), owner, name)
+	if err != nil {
+		return errors.Wrap(err, "Unable to get repository permissions"), http.StatusNotFound
+	}
+
+	if !(repo.ViewerPermission == "ADMIN" || repo.ViewerPermission == "MAINTAIN" || repo.ViewerPermission == "WRITE") {
```
Comment (tsenart) [inline enterprise/pkg/codeintel/lsifserver/proxy/auth.go]:
```suggestion
        switch repo.ViewerPermission {
        case "ADMIN", "MAINTAIN", "WRITE":
            return nil, 0
        default:
            return errors.New("You do not have write permission to the repository."), http.StatusUnauthorized
        }
```

File: enterprise/pkg/codeintel/lsifserver/proxy/upload.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	"github.com/sourcegraph/sourcegraph/internal/conf"
+)
+
+func uploadProxyHandler(p *httputil.ReverseProxy) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		q := r.URL.Query()
+		repoName := q.Get("repository")
+		commit := q.Get("commit")
+
+		repo, err := backend.Repos.GetByName(r.Context(), api.RepoName(repoName))
```
Comment (tsenart) [inline enterprise/pkg/codeintel/lsifserver/proxy/upload.go]: FYI: This has an implicit dependency on `dbconn.Global` being instantiated, which happens at the beginning of the frontend main function: https://github.com/sourcegraph/sourcegraph/blob/master/enterprise/cmd/frontend/main.go#L37

PS: One day, we'll start moving towards a world without such hard to test global state magic. https://peter.bourgon.org/blog/2017/06/09/theory-of-modern-go.html

Comment (efritz) [reply enterprise/pkg/codeintel/lsifserver/proxy/upload.go]: Should there be a change (code or docs) here?

Related: I've used [service injection](https://github.com/go-nacelle/service) to great success in past projects.

File: cmd/frontend/internal/httpapi/httpapi.go (hunk 1 above)
Comment (efritz) [reply cmd/frontend/internal/httpapi/httpapi.go]: Sure thing, I'll update this now.

File: enterprise/pkg/codeintel/lsifserver/proxy/upload.go (hunk 3 above)
Comment (tsenart) [reply enterprise/pkg/codeintel/lsifserver/proxy/upload.go]: > Should there be a change (code or docs) here?

I don't think so. This is how things are done in most of our code base for now (but not everywhere).

> Related: I've used service injection to great success in past projects.

❤️

**Comment by mrnugget** [comment]
@efritz If you had ideas on how to improve this "pattern" (define interfaces in frontend, inject concrete enterprise implementations at runtime, etc.), let's chat! Because I'm not sure I really like it.

**Pull request merged**